hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/schema/MemberSchemaService.java

```diff
@@ -16,113 +16,85 @@
 import java.util.Map;
 import java.util.Properties;
-import java.util.Set;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
 
-import static com.hazelcast.internal.util.InvocationUtil.invokeOnStableClusterSerial;
-
-public class MemberSchemaService implements ManagedService, PreJoinAwareService, SchemaService {
+/**
+ * Responsible for replicating the schemas across the cluster and giving a
+ * mechanism for clients/members to access schemas that they don't own in their
+ * local registry.
+ */
+public class MemberSchemaService implements
+        ManagedService,
+        PreJoinAwareService,
+        SchemaService,
+        SplitBrainHandlerService,
+        CoreService {
 
-    private static final int MAX_RETRIES = 100;
     private final Map<Long, Schema> schemas = new ConcurrentHashMap<>();
+    private final SchemaReplicator replicator = new SchemaReplicator(this);
+    private final CompactSchemaExtractor schemaExtractor = new CompactSchemaExtractor(this);
+
     private ILogger logger;
     private NodeEngine nodeEngine;
 
     @Override
     public void init(NodeEngine nodeEngine, Properties properties) {
         this.logger = nodeEngine.getLogger(SchemaService.class);
         this.nodeEngine = nodeEngine;
+
+        // Set the node engine reference to replicator as well.
+        replicator.init(nodeEngine);
+
+        boolean useBigEndian = nodeEngine.getConfig()
+                .getSerializationConfig()
+                .getByteOrder() == ByteOrder.BIG_ENDIAN;
+        schemaExtractor.setUseBigEndian(useBigEndian);
     }
 
     @Override
     public void reset() {
+        // Called in the current node before it joins to larger cluster in the
+        // split-brain healing process. Resetting the state of the replications
+        // allow replaying the replications in the larger cluster in this node.
+        // Before calling this, we already copied current state of the
+        // replications so that the replications occurred in the smaller
+        // cluster can be replicated to the larger cluster as well.
+        // See `prepareMergeRunnable` for that.
+        replicator.clear();
     }
 
     @Override
     public void shutdown(boolean terminate) {
         schemas.clear();
+        replicator.clear();
     }
 
-    @Override
-    public Operation getPreJoinOperation() {
-        if (schemas.size() == 0) {
-            return null;
-        }
-        if (logger.isFinestEnabled()) {
-            logger.finest("Preparing prejoin operation with schemas " + schemas);
-        }
-        return new SendAllSchemasOperation(new ArrayList<>(schemas.values()));
-    }
-
+    @Nullable
     @Override
     public Schema get(long schemaId) {
-        return getAsync(schemaId).join();
-    }
-
-    public CompletableFuture<Schema> getAsync(long schemaId) {
         if (!nodeEngine.getClusterService().getClusterVersion().isEqualTo(Versions.V5_2)) {
```

Comment:
```suggestion
        if (nodeEngine.getClusterService().getClusterVersion().isLessThan(Versions.V5_2)) {
```

Reply:
I will remove this check altogether after merging this PR, with other BETA related tags, documentations etc.